mayastor/src/core/descriptor.rs

```diff
@@ -109,3 +166,41 @@ impl Debug for Descriptor {
         )
     }
 }
+
+extern "C" fn spdk_range_cb(
+    ctx: *mut ::std::os::raw::c_void,
+    status: ::std::os::raw::c_int,
+) {
+    unsafe {
+        let s = ctx as *mut mpsc::Sender<i32>;
+        if let Err(e) = (*s).start_send(status) {
+            panic!("Failed to send SPDK completion with error {}.", e);
+        }
+    }
+}
+
+/// Store the context for calls to lock_lba_range and unlock_lba_range.
+/// Corresponding lock/unlock calls require the same context to be used.
+pub struct RangeContext {
+    pub offset: u64,
+    pub len: u64,
+    io_channel: Arc<IoChannel>,
```

Comment:
Everything that is saved in the context is there to ensure that the parameters passed to the bdev_lock_lba_range function match those passed to the bdev_unlock_lba_range. 

The unlock expects all the parameters to match the corresponding lock; if not it returns an error.

Reply:
I used Arc<> here because multiple rebuild jobs can be running concurrently and using the same channel. This was primarily to prevent a channel having to be created each time a rebuild job runs.

Reply:
I think what it asserted is that the pointer address is still the same, and it is not deep comparing bit for bit that what there during lock is there for unlock? 

If you hold onto the RangeContext, the fact that you hold onto it as you need it later implies that it is not dropped. The Arc<> keeps a reference count (atomic) to the channel, such that when nobody is using it its get dropped automatically (i.e garbage collected). 

Also, Arc<> implies that the channel is used by multiple different cores, we should never share channels across cores.